File: services/web/src/auth/internalJwt.ts
import jwt from "jsonwebtoken";

// Must match CloudCure.Api.Auth.InternalJwt on the .NET side exactly.
export const INTERNAL_JWT_ISSUER = "cloudcure-web";
export const INTERNAL_JWT_AUDIENCE = "cloudcure-api";

/**
 * Mints a fresh, short-lived internal service-to-service JWT for one outgoing API request.
 * Short-lived by design (60s) so a token leaked from a log/container has a tiny window of
 * use, unlike a long-lived credential. Never Auth0's own tokens — this is Node vouching for
 * a caller (or, before login resolves an identity, for itself) to the internal API.
 */
export function mintInternalToken(personId?: string, roles: string[] = []): string {
  const secret = process.env.INTERNAL_JWT_SECRET;
  if (!secret) {
    throw new Error("Missing required environment variable: INTERNAL_JWT_SECRET");
  }

  const claims: Record<string, unknown> = {};
  if (personId) {
    claims.sub = personId;
  }
  if (roles.length > 0) {
    claims.role = roles;
  }

  return jwt.sign(claims, secret, {
    issuer: INTERNAL_JWT_ISSUER,
    audience: INTERNAL_JWT_AUDIENCE,
    expiresIn: "60s",
  });
}

File: services/api/src/CloudCure.Api/Auth/InternalJwt.cs
namespace CloudCure.Api.Auth;

/// <summary>
/// Constants for the internal, service-to-service HS256 JWT the Node BFF mints on every
/// outgoing request (see plan section 3). Never Auth0's own tokens — Auth0 is identity-only
/// here, and Node is the only thing that ever talks to it.
/// </summary>
public static class InternalJwt
{
    public const string Issuer = "cloudcure-web";
    public const string Audience = "cloudcure-api";
    public const string RoleClaimType = "role";
    public const string SubjectClaimType = "sub";
}
